counsel.el

```diff
@@ -4564,6 +4564,14 @@ Note: Duplicate elements of `kill-ring' are always deleted."
  '(("d" counsel-yank-pop-action-remove "delete")
    ("r" counsel-yank-pop-action-rotate "rotate")))
 
+(defcustom counsel-yank-delete-selection (when delete-selection-mode 'yank)
+  "Control if `counsel-yank-pop' deletes the active region before yanking.
+This setting rely on the `delete-selection-mode'."
+  :type 'symbolp)
+
+(when counsel-yank-delete-selection
+  (put #'counsel-yank-pop 'delete-selection counsel-yank-delete-selection))
```

Comment:
It's better to minimise side-effects at load-time.  I think you should put:

```el
(put #'counsel-yank-pop 'delete-selection 'yank)
```

right after `(defun counsel-yank-pop ...)`.

If `delete-selection-mode` is disabled, the symbol property is a no-op anyway.